src/Data/Array.purs

```diff
@@ -1083,8 +1086,33 @@ union = unionBy (==)
 -- | unionBy mod3eq [1, 5, 1, 2] [3, 4, 3, 3] = [1, 5, 1, 2, 3]
 -- | ```
 -- |
-unionBy :: forall a. (a -> a -> Boolean) -> Array a -> Array a -> Array a
-unionBy eq xs ys = xs <> foldl (flip (deleteBy eq)) (nubByEq eq ys) xs
+unionBy :: forall a. (a -> a -> Ordering) -> Array a -> Array a -> Array a
+unionBy cmp left right = map snd $ ST.run do
```

Comment:
It may be more efficient to take advantage of the of the fact that the output is always the first array plus some other stuff appended to it. Here's some non-ST pseudo-ish code describing that option:
```purs
union left right = left <> other where
  other =
    combineIndex right left -- intentionally putting right first and assuming stable sort
    # sortBy valueFunc
    # groupBy valueFunc
    # map head
    # filter keepRightOnly
```

Reply:
What would be the runtime cost of this approach versus mine?

I made the rather stupid assumption that mine would be `n+m` because when Harry said the current version was `n*m`, I thought he was implying we should make the code linear. I didn't actually analyze my code using Big O notation to see how many steps it takes.

Reply:
I believe all reasonable approaches are `O(n log n)` (where n is the sum of array sizes). But benchmarking could still reveal a 3x speedup with a different approach, which would still be `O(n log n)`. I'll play around with this and report back with findings.